src/RelayClient.ts

```diff
@@ -38,57 +44,83 @@ import { isDeployRequest } from './common/relayRequest.utils';
 import type { EnvelopingTxRequest } from './common/relayTransaction.types';
 import {
   MISSING_CALL_FORWARDER,
-  MISSING_SMART_WALLET_ADDRESS,
   NOT_RELAYED_TRANSACTION,
 } from './constants/errorMessages';
 import EnvelopingEventEmitter, {
   envelopingEvents,
 } from './events/EnvelopingEventEmitter';
 import { estimateRelayMaxPossibleGas } from './gasEstimator';
 import {
-  applyGasCorrectionFactor,
-  applyInternalEstimationCorrection,
-  getEnvelopingConfig,
+  estimateInternalCallGas,
+  estimateTokenTransferGas,
+  getSmartWalletAddress,
   selectNextRelay,
   validateRelayResponse,
 } from './utils';
 
-class RelayClient extends EnvelopingEventEmitter {
-  private readonly _provider: providers.Provider;
+const isNullOrUndefined = (value: unknown) => value === null || value === undefined;
 
+class RelayClient extends EnvelopingEventEmitter {
   private readonly _envelopingConfig: EnvelopingConfig;
 
   private readonly _httpClient: HttpClient;
 
   constructor() {
     super();
 
-    this._provider = getDefaultProvider();
     this._envelopingConfig = getEnvelopingConfig();
     this._httpClient = new HttpClient();
   }
 
   private _getEnvelopingRequestDetails = async (
-    envelopingRequest: UserDefinedEnvelopingRequest,
-    {
-      forceGasLimit,
-      forceGasPrice,
-    }: RequestConfig
+    envelopingRequest: UserDefinedEnvelopingRequest
   ): Promise<EnvelopingRequest> => {
     const isDeployment: boolean = isDeployRequest(
       envelopingRequest as EnvelopingRequest
     );
     const {
-      relayData: { callForwarder },
-      request: { data, from, to, tokenContract },
+      relayData,
+      request,
     } = envelopingRequest;
 
-    if (!callForwarder) {
-      throw new Error(MISSING_CALL_FORWARDER);
+    const { from, tokenContract } = request;
+
+    if (!from) {
+      throw new Error('Field `from` is not defined in request body.');
+    }
+
+    if (!tokenContract) {
+      throw new Error('Field `tokenContract` is not defined in request body.');
+    }
+
+    if (!isDeployment) {
+      if (!relayData?.callForwarder) {
+        throw new Error(MISSING_CALL_FORWARDER);
```

Comment:
Why do we store only this error message in the constant? I'd try to be consistent and do the same with the others.

Reply:
updated